kong/db/declarative/import.lua

```diff
@@ -294,7 +295,12 @@ local function load_into_cache(entities, meta, hash)
 
       if schema.cache_key then
         local cache_key = dao:cache_key(item)
-        t:set(cache_key, item_marshalled)
+        if #cache_key > LMDB_MAXKEYSIZE then
+          local hash_cache_key = sha256(cache_key)
```

Comment:
It may not be the correct place.
If we change the cache key here, we will not find it with the original cache_key.

should we take a look at `dao:cache_key`

Reply:
Yes, we cannot just start hashing cache keys in a single place. Either it has to be everywhere (which is too risky), or you need to create a custom dao for key_auth (https://github.com/Kong/kong/blob/master/kong/plugins/key-auth/daos.lua) -> a `dao` property. see: https://github.com/Kong/kong/blob/master/kong/plugins/oauth2/daos.lua#L92 and https://github.com/Kong/kong/blob/master/kong/plugins/oauth2/daos/oauth2_tokens.lua#L8-L10.

Reply:
> Either it has to be everywhere (which is too risky), or you need to create a custom dao for key_auth

It looks like we need to do the length check at the only place where the cache_key is generated.

When loading the configuration into the cache (lmdb), the cache_key needs to be generated, where the key is directly the table, and the content:

```json
{
  consumer = {
    id = "075415e8-e1cf-546e-8f8c-b7e15251e3d7"
  },
  created_at = 1676408487,
  id = "3d9ec867-3822-5729-adbf-775e972cbb74",
  key = "6y8fivuu6jj822b77hcuxp6z33pxgie3t5ypx8mwuxne59gp3yvgt62m2vazuu8b3ahyjhzdmri8bk768qpdp745ijt362f3z5nnwgn785iakpwia5pegwpxki4k8xjzrac2nbg2ff3adrbbgmw7ihj8mj4itwmjaxkkrany5qnb7za46imhc9vwqmg4vpqf5vkza5dfgkr8fphpfrk6bcgz2mxuqgvcvadjb974jpjuctfjptw4j5izb68pywg9a7h85wfdddxh7j7j9yd2cxtvyk7n4cggibec9qtjhci8rdhuwcf2ixtfk4vm6hzj5j6ece9xwuxicdtrizwgm6mrx99av4ukw4c9qp7yr23ij9tutph8gek698765ydjrmgkm88pyzuq7jabpeqe7ae9j8qiew2d9rjdmmd27nmtkdgajrg78347dmrbgvxzjw3f4ewakckxnia3izv7itqa9bufnjn7rd9mwkvzfmuba4hrwk49f2ktzw9npw94",
  ws_id = "0dc6f45b-8f8d-40d2-a504-473544ee190b"
}
```

stack traceback:
```
	./kong/db/dao/init.lua:1539: in function 'cache_key'
	./kong/db/declarative/import.lua:300: in function 'load_into_cache'
	./kong/init.lua:516: in function <./kong/init.lua:511>
	[C]: in function 'pcall'
	./kong/concurrency.lua:54: in function 'with_worker_mutex'
	./kong/init.lua:511: in function 'load_declarative_config'
	./kong/init.lua:763: in function 'init_worker'
	init_worker_by_lua:2: in main chunk, context: init_worker_by_lua*
```

When querying the cache, we also need to generate the cache_key, where the key is a string, 

stack traceback:
```
	./kong/db/dao/init.lua:1498: in function 'cache_key'
	./kong/plugins/key-auth/handler.lua:177: in function 'do_authentication'
	./kong/plugins/key-auth/handler.lua:227: in function <./kong/plugins/key-auth/handler.lua:215>
```

This seems like a good choice for the `DAO:cache_key` function to avoid making changes to potentially long cache_key that are distributed across different plugins.